packages/react/select/src/Select.tsx

```diff
@@ -990,36 +971,196 @@ const SelectPopperPosition = React.forwardRef<
+   * Can be prevented.
+   */
+  onPointerDownOutside?: DismissableLayerProps['onPointerDownOutside'];
+}
+const SelectContentModalityImpl = React.forwardRef<
+  SelectContentModalityImplElement,
+  SelectContentModalityImplProps
+>((props: ScopedProps<SelectContentModalityImplProps>, forwardedRef) => {
+  const {
+    __scopeSelect,
+    disableOutsideScroll,
+    trapFocus,
+    disableOutsidePointerEvents,
+    onCloseAutoFocus,
+    onEscapeKeyDown,
+    onPointerDownOutside,
+    onFocusOutside,
+    onInteractOutside,
+    onDismiss,
+    children,
+  } = props;
+  const context = useSelectContext(CONTENT_NAME, __scopeSelect);
+
+  return (
+    <RemoveScroll as={Slot} allowPinchZoom enabled={disableOutsideScroll}>
+      <FocusScope
+        asChild
+        trapped={trapFocus}
+        onMountAutoFocus={(event) => {
+          // we prevent open autofocus because we manually focus the selected item
+          event.preventDefault();
+        }}
+        onUnmountAutoFocus={composeEventHandlers(onCloseAutoFocus, (event) => {
+          context.trigger?.focus({ preventScroll: true });
+          event.preventDefault();
+        })}
+      >
+        <DismissableLayer
+          ref={forwardedRef}
+          asChild
+          disableOutsidePointerEvents={disableOutsidePointerEvents}
+          onEscapeKeyDown={onEscapeKeyDown}
+          onPointerDownOutside={onPointerDownOutside}
+          onFocusOutside={onFocusOutside}
+          onInteractOutside={onInteractOutside}
+          onDismiss={onDismiss}
+        >
+          {children}
+        </DismissableLayer>
+      </FocusScope>
+    </RemoveScroll>
+  );
+});
+
+SelectContentModalityImpl.displayName = CONTENT_MODALITY_IMPL_NAME;
+
+type SelectContentModalityImplPublicProps = Omit<
+  SelectContentModalityImplProps,
+  keyof SelectContentModalityImplPrivateProps
+>;
+interface SelectContentModalProps extends SelectContentModalityImplPublicProps {}
+const SelectContentModal = (props: ScopedProps<SelectContentModalProps>) => {
+  const { __scopeSelect, children, ...modalityProps } = props;
+  const context = useSelectContext(CONTENT_NAME, __scopeSelect);
+  const contentRef = React.useRef<SelectContentModalityImplElement>(null);
+
+  // aria-hide everything except the content (better supported equivalent to setting aria-modal)
+  React.useEffect(() => {
+    const content = contentRef.current;
+    if (content) return hideOthers(content);
+  }, []);
+
+  return (
+    <SelectContentModalityImpl
+      ref={contentRef}
+      // we make sure we're not trapping once it's been closed
+      // (closed !== unmounted when animating out)
+      trapFocus={context.open}
+      // make sure to only disable pointer events when open
+      // this avoids blocking interactions while animating out
+      disableOutsidePointerEvents={context.open}
+      disableOutsideScroll
+      {...modalityProps}
+      onFocusOutside={(event) => event.preventDefault()}
```

Comment:
I have seen a problem on mobile in non-modal mode where the `Trigger` gets focus after being opened. Maybe I need to put this line in both modal and non-modal mode? 

I noticed that this same issue also occurs in the dropdown menu, I believe it is because of react 17? In the documentation it doesn't seem to have any issues :thinking: